### ISSUE TITLE
More information on Fix ERROR on ligatures conditional #1596

### ISSUE THREAD

**Issue by kontur** [googlefonts/fontbakery#3047]
## Observed behaviour

I am running a check with `-c com.google.fonts/check/ligature_carets` and get the error message:

>  * FAIL: Failed to lookup ligatures. This font file seems to be malformed. For more info, read: https://github.com/googlefonts/fontbakery/issues/1596 [code: malformed]

## Expected behaviour

Getting a better error message, or not failing. No error message is returned, and when inspecting what exception is raised from fontTools it is an ValueError. With -v this information should, at the very least, be returned for debugging.

## Resources and exact process needed to replicate

The font I am working with is proprietary so I cannot share, but unlike specified in #1596 I can successfully ttx that font. Also, ligatures seem to work. The font has been subset with `pyftsubset`, if that gives any clues.

When investigating what causes the fontTools error for which Fontbakery consequently bails I found that in the `ligature_glyphs` condition the problem seems to be encountering an Extension subtable (for which I believe there is not supposed to be a `ligature` attribute, which is causing the attribute access error and fails checking all subsequent tables).

Changing [this line](https://github.com/googlefonts/fontbakery/blob/84fe75e0d9b541302a2d83fe7b20d49df0799058/Lib/fontbakery/profiles/shared_conditions.py#L121) to:

```
if subtable.Format == 1 and not isinstance(subtable, fontTools.ttLib.tables.otTables.ExtensionSubst):
…
```

results in the check correctly running. It then warns me about missing caret positions, which I interpret as the check itself actually running correctly. I am not 100% if my assumption here is correct about Extension tables, but if this is the reason some odd fonts fail it would be good to double check if with those fonts, too, the issue is this and not something else. If the assumption is correct, the internal error can be avoided.

**Comment by felipesanches** [comment]
When you say "error" here in this issue I think you may actually mean what we call a "fail", right?

In FontBakery jargon we refer to **ERROR**s when something crashes due to a bug or to an incorrect installation of the software itself or its dependencies. And when we talk about **FAIL**s, we're referring to problems in a font file.

**Comment by felipesanches** [comment]
well... maybe you're really talking about an ERROR though. I'll look more carefully

**Comment by kontur** [comment]
Yes and no. I think it is in fact an _ERROR_ that prevents the check from running like it should, for this particular font. Encountering the unexpected subtable format makes the `ligature_glyphs` method catch an error, for which the method inconveniently returns a `-1` — which then "neatly" _FAIL_s the font in the context of the check run.

**Comment by kontur** [comment]
That said, I am not opposing the `try` block here. I seen in the related issue that it is not entirely clear what causes the problem. But: Since I have one specific case where I can narrow down in more detail what causes the problem, I figure I suggest a more precise `if` clause that will at least not raise an error for this particular case (possibly those other cases as well).

**Comment by khaledhosny** [comment]
The code currently also assumes `liga` feature will always have type 4 lookups, which is a requirement and any lookup type can be potentially used here and the font would still be legitimate.

The code should check the lookup type before assuming it is a ligature lookup:
https://github.com/googlefonts/fontbakery/blob/84fe75e0d9b541302a2d83fe7b20d49df0799058/Lib/fontbakery/profiles/shared_conditions.py#L139-L142

**Comment by kontur** [comment]
Returning to this same error with another odd observation. I've tried this with a fresh install of Fontbakery (0.7.34). I seem to be getting an exception for _any_ font I check, the exception at its root being the `if subtable.Format == 1` check, where `Format` is not an attribute on subtable. I'll return with more info when I'm don scratching my head.

**Comment by kontur** [comment]
This seems to be a regression introduced between fonttools 4.22.0 and 4.21.0.

With _any_ font, but e.g. [Noto Sans](https://github.com/googlefonts/noto-fonts/blob/master/hinted/ttf/NotoSans/NotoSans-Black.ttf):

With fontbakery==0.7.34 && fontTools==4.22.0 and this print statement added:

```
try:
        if "GSUB" in ttFont and ttFont["GSUB"].table.LookupList:
            for record in ttFont["GSUB"].table.FeatureList.FeatureRecord:
                if record.FeatureTag == 'liga':
                    for index in record.Feature.LookupListIndex:
                        lookup = ttFont["GSUB"].table.LookupList.Lookup[index]
                        for subtable in lookup.SubTable:
                            print("SUBTABLE", hasattr(subtable, "Format"))
                            if subtable.Format == 1:
                                for firstGlyph in subtable.ligatures.keys():
                                    for lig in subtable.ligatures[firstGlyph]:
                                        if lig.LigGlyph not in all_ligature_glyphs:
                                            all_ligature_glyphs.append(lig.LigGlyph)
        return all_ligature_glyphs
    except:
        return -1  # Indicate fontTools-related crash...
```

The output for: `fontbakery check-googlefonts -c com.google.fonts/check/ligature_carets ~/Desktop/NotoSans-Black.ttf -n` is:

> Start ... running 1 individual check executions.
SUBTABLE False
 >> com.google.fonts/check/ligature_carets
    Are there caret positions declared for every ligature?
    with /Users/johannes/Desktop/NotoSans-Black.ttf

      Rationale:                                                                
      All ligatures in a font must have corresponding caret (text cursor)       
      positions defined in the GDEF table, otherwhise, users may experience     
      issues with caret rendering.                                              
                                                                                
      If using GlyphsApp, ligature carets can be set directly on canvas by      
      accessing the `Glyph -> Set Anchors` menu option or by pressing the       
      `Cmd+U` keyboard shortcut.                                                
                                                                                
      If designing with UFOs, (as of Oct 2020) ligature carets are not yet      
      compiled by ufo2ft, and therefore will not build via FontMake. See        
      googlefonts/ufo2ft/issues/329                                             

    FAIL Failed to lookup ligatures. This font file seems to be malformed. For more
info, read: https://github.com/googlefonts/fontbakery/issues/1596      
         [code: malformed]                                                      


    Result: FAIL


But with `pip install fontTools==4.21.0` the same command returns:

> Start ... running 1 individual check executions.
SUBTABLE True

Total:

    ERROR: 0
    FAIL: 0
    WARN: 0
    INFO: 0
    SKIP: 0
    PASS: 1


DONE!

**Comment by kontur** [comment]
Ah... from the fontTools 0.4.22 release notes:

```
4.22.0

@madig madig released this yesterday · 1 commit to main since this release

    [ttLib] Remove .Format from Coverage, ClassDef, SingleSubst, LigatureSubst, AlternateSubst, MultipleSubst (#2238).
    ATTENTION: This will change your TTX dumps!
```

Alas, this seems intentional, but then the check needs to be updated.
